File: 475. Heaters/solution.go
import "sort"

func findRadius(houses []int, heaters []int) (max int) {
	sort.Ints(houses)
	sort.Ints(heaters)
	for i, j := 0, 0; i < len(houses); i++ {
		if houses[i] > heaters[j] {
			for j+1 < len(heaters) && abs(houses[i]-heaters[j]) >= abs(houses[i]-heaters[j+1]) {
				j++
			}
		}

		d := abs(houses[i] - heaters[j])
		if max < d {
			max = d
		}
	}
	return
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: 475. Heaters/solution.js
/**
 * @param {number[]} houses
 * @param {number[]} heaters
 * @return {number}
 */
var findRadius = function(houses, heaters) {
  let cur = 0;
  let next = 0;
  let radius = 0;

  houses.sort((a, b) => a - b);
  heaters.sort((a, b) => a - b);

  for (let i = 0; i < houses.length; i++) {
    while (next <= heaters.length - 1 && houses[i] > heaters[next]) {
      cur = next;
      next++;
    }

    if (next > heaters.length - 1) {
      next = cur;
    }

    if (cur === next) {
      radius = Math.max(radius, Math.abs(heaters[cur] - houses[i]));
    } else {
      radius = Math.max(
        radius,
        Math.min(heaters[next] - houses[i], houses[i] - heaters[cur])
      );
    }
  }

  return radius;
};
